pkg/releaseutil/kind_sorter.go

```diff
@@ -111,48 +111,64 @@ var UninstallOrder KindSortOrder = []string{
 //
 // Results are sorted by 'ordering', keeping order of items with equal kind/priority
 func sortManifestsByKind(manifests []Manifest, ordering KindSortOrder) []Manifest {
+	lessFunc := lessByKind(ordering)
 	sort.SliceStable(manifests, func(i, j int) bool {
-		return lessByKind(manifests[i], manifests[j], manifests[i].Head.Kind, manifests[j].Head.Kind, ordering)
+		return lessFunc(manifests[i].Head.Kind, manifests[j].Head.Kind)
 	})
 
 	return manifests
 }
 
-// sort hooks by kind, using an out-of-place sort to preserve the input parameters.
-//
-// Results are sorted by 'ordering', keeping order of items with equal kind/priority
-func sortHooksByKind(hooks []*release.Hook, ordering KindSortOrder) []*release.Hook {
+// sortHooks sorts hooks by weight, kind, and finally by name.
+// Kind order is defined by ordering.
+func sortHooks(hooks []*release.Hook, ordering KindSortOrder) []*release.Hook {
 	h := hooks
+
+	// Sort first by name, the least important ordering.
+	sort.Slice(h, func(i, j int) bool {
+		return h[i].Name < h[j].Name
+	})
+
+	// Then sort by kind, keeping equal elements in their original order (Stable).
+	lessFunc := lessByKind(ordering)
+	sort.SliceStable(h, func(i, j int) bool {
+		return lessFunc(h[i].Kind, h[j].Kind)
+	})
+
+	// Finally, sort by weight, again keeping equal elements in their original order.
 	sort.SliceStable(h, func(i, j int) bool {
-		return lessByKind(h[i], h[j], h[i].Kind, h[j].Kind, ordering)
+		return h[i].Weight < h[j].Weight
 	})
 
 	return h
 }
 
-func lessByKind(a interface{}, b interface{}, kindA string, kindB string, o KindSortOrder) bool {
+// lessByKind takes a KindSortOrder and returns a LessFunc that compares two kinds according to said order.
+func lessByKind(o KindSortOrder) func(kindA string, kindB string) bool {
```

Comment:
I imagine that we return a map because we want to do slice->map only once.

I wonder if the increased complexity is balanced by the performance improvement

I would add a comment or in the commit message some info about the choice